java/code/src/com/redhat/rhn/taskomatic/task/MinionActionExecutor.java

```diff
@@ -192,4 +201,13 @@ private long countQueuedServerActions(Action action) {
                      .filter(serverAction -> ActionFactory.STATUS_QUEUED.equals(serverAction.getStatus()))
                      .count();
     }
+
+    private boolean allServerActionsFinished(Action action) {
+        return action != null &&
+            !CollectionUtils.isEmpty(action.getServerActions()) &&
```

Comment:
does this mean an action with no server actions does never count as having all server actions finished?

Reply:
i guess it makes sense given the `countQueuedServerActions(action) == 0` check already. I'm not exactly sure in what scenario an action would be saved and scheduled without server actions and then in a separate transaction server actions get added. It sounds like all kinds of problematic.

Reply:
i think its fine for now. The code suggests that there are cases where the action has no server actions initially. The whole action hand off between tomcat and taskomatic seems just very fragile so i think this needs some proper way in the future.